File: 547_Number_of_Provinces/547_1.py
from typing import List

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parents = [i for i in range(n)]
        sizes = [1 for i in range(n)]
        count = n

        def find(x):
            while x != parents[x]:
                x = parents[x]
                parents[x] = parents[parents[x]]
            return x
        
        def union(i, j):
            nonlocal count
            root1 = find(i)
            root2 = find(j)
            if root1 == root2:
                return
            if sizes[root1] < sizes[root2]:
                parents[root1] = root2
                sizes[root2] += sizes[root1]
            else:
                parents[root2] = root1
                sizes[root1] += sizes[root2]
            count -= 1

        for i in range(n):
            for j in range(i + 1, n):
                if isConnected[i][j] == 1:
                    union(i, j)
        return count

File: 547_Number_of_Provinces/547_1.c
void dfs(int **isConnected, int i, int n, bool visited[n]) {
    for (int j = 0; j < n; j++) {
        if (isConnected[i][j] == 1 && !visited[j]) {
            visited[j] = true;
            dfs(isConnected, j, n, visited);
        }
    }
}

int findCircleNum(int** isConnected, int isConnectedSize, int* isConnectedColSize){
    int n = isConnectedSize, count = 0;
    bool visited[n];
    memset(visited, false, sizeof(visited));
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            count++;
            visited[i] = true;
            dfs(isConnected, i, n, visited);
        }
    }
    return count;
}
